### PR TITLE
Wsjtx

### PR DESCRIPTION
**Pull request by xylo04**
### Please confirm your submission meets all the criteria

- [x] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [x] My pull request follows the instructions at [App Submission][submission].
- [ ] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)* 
  **App needs access to windowing; devices and sound for amateur radio control; home directory for standard file locations; network for internet service integration**
- [x] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
- [x] I am an upstream contributor to the project. If not, I contacted upstream developers about submitting their software to Flathub. **[Link](https://sourceforge.net/p/wsjt/mailman/wsjt-devel/thread/CAAY2tUj1XX%3DoKfjK_fJ3LkMzGu9Zt6Rb-CAvcZSEXhBKtN9Aew%40mail.gmail.com/#msg37311785)**
- [ ] I own the domain used in the application ID or the domain has a policy for delegating subdomains (e.g. GitHub, SourceForge).
  **I don't own the domain, but upstream does**
- [ ] Any additional patches or files have been submitted to the upstream projects concerned. *(If not, explain why.)*
  **One patch required to change icon install location**

@AsciiWolf has offered to co-maintain

[reqs]: https://github.com/flathub/flathub/wiki/App-Requirements
[maint]: https://github.com/flathub/flathub/wiki/App-Maintenance
[submission]: https://github.com/flathub/flathub/wiki/App-Submission

### CODE REVIEW

File: edu.princeton.physics.WSJTX.metainfo.xml [hunk 1]
```diff
@@ -0,0 +1,92 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<component type="desktop">
+  <id>edu.princeton.physics.WSJTX.desktop</id>
```
Comment (hfiguiere) [inline edu.princeton.physics.WSJTX.metainfo.xml:3]:
```suggestion
  <id>edu.princeton.physics.WSJTX</id>
```

**Comment by hfiguiere** [comment]
bot, build edu.princeton.physics.WSJTX

File: edu.princeton.physics.WSJTX.yml [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+  - /man
+  - /include
+  - /share/man
+  - /share/doc
+  - /doc
+  - '*.a'
+  - '*.la'
+modules:
+  - shared-modules/libusb/libusb.json
+  - hamlib.yml
```
Comment (hfiguiere) [inline edu.princeton.physics.WSJTX.yml:26]: it's not common practice to simply split modules like that.

File: fftw3.yml [hunk 3]
```diff
@@ -0,0 +1,15 @@
+name: fftw3
```
Comment (hfiguiere) [inline fftw3.yml:1]: fftw3f is in shared modules, and it build with the SIMD for the supported platforms. (they are opt-in not opt-out)

**Review by hfiguiere — CHANGES_REQUESTED** [review]

File: edu.princeton.physics.WSJTX.metainfo.xml [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+  <name>wsjtx</name>
+  <summary>Amateur Radio Weak Signal Operating</summary>
+  <developer_name>Joe Taylor and others</developer_name>
+  <description>
+  <p>
+WSJT-X Version 2.1 offers ten different protocols or modes: FT4, FT8,
```
Comment (AsciiWolf) [inline edu.princeton.physics.WSJTX.metainfo.xml:9]: Here is a description from Debian that we could use:

> WSJT-X implements amateur radio communication protocols or "modes" called FT4, FT8, JT4, JT9, JT65, QRA64, ISCAT, MSK144, and WSPR, as well as one called Echo for detecting and measuring your own radio signals reflected from the Moon. These modes were all designed for making reliable, confirmed QSOs under extreme weak-signal conditions.
> 
> JT4, JT9, JT65, and QRA64 use nearly identical message structure and source encoding (the efficient compression of standard messages used for minimal QSOs). They use timed 60-second T/R sequences synchronized with UTC. JT65 and QRA64 were designed for EME ("moonbounce") on the VHF/UHF bands; JT65 has also proved popular and effective for worldwide QRP communication at HF. JT9 is optimized for the LF, MF, and HF bands. It is about 2 dB more sensitive than JT65 while using less than 10% of the bandwidth. With either JT9 or JT65, world-wide QSOs are possible with power levels of a few watts and compromise antennas. JT4 and QRA64 are optimized for EME on the VHF and higher bands, and especially the microwave bands from 2.3 to 24 GHz.
> 
> FT4 and FT8 are operationally similar but use T/R cycles only 7.5 and 15 s long, respectively. MSK144 is designed for Meteor Scatter on the VHF bands. These modes offer enhanced message formats with support for nonstandard callsigns and some popular contests.
> 
> WSPR mode implements a protocol designed for probing potential propagation paths with low-power transmissions. WSPR is fully implemented within WSJT-X, including programmable "band-hopping".

File: edu.princeton.physics.WSJTX.metainfo.xml [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+</description>
+  <metadata_license>CC0-1.0</metadata_license>
+  <project_license>GPL-3.0+</project_license>
+  <url type="bugtracker">https://sourceforge.net/projects/wsjt/lists/wsjt-devel</url>
+  <url type="homepage">https://physics.princeton.edu/pulsar/K1JT/index.html</url>
+  <project_group>none</project_group>
+  <screenshots>
+    <screenshot type="default">
+      <caption>Main WSJT-X window</caption>
+      <image height="743" width="774">https://sourceforge.net/p/wsjt/wsjtx/ci/master/tree/doc/user_guide/en/images/main-ui.png?format=raw</image>
```
Comment (AsciiWolf) [inline edu.princeton.physics.WSJTX.metainfo.xml:84]:
```suggestion
      <image>https://sourceforge.net/p/wsjt/wsjtx/ci/master/tree/doc/user_guide/en/images/main-ui.png?format=raw</image>
```

**Comment by AsciiWolf** [comment]
@xylo04 Here is a fix for both the AppStream metadata issues: https://github.com/xylo04/flathub/pull/1

File: edu.princeton.physics.WSJTX.metainfo.xml (hunk 1 above)
Comment (AsciiWolf) [reply edu.princeton.physics.WSJTX.metainfo.xml:3]: Hmm, yeah. The desktop suffix is not needed.

File: fftw3.yml (hunk 3 above)
Comment (AsciiWolf) [reply fftw3.yml:1]: Thanks, you are right! Here is it: https://github.com/flathub/shared-modules/blob/master/linux-audio/fftw3f.json

File: boost.yml [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+    url: https://boostorg.jfrog.io/artifactory/main/release/1.74.0/source/boost_1_74_0.tar.bz2
+    sha256: 83bfc1507731a0906e387fc28b7ef5417d591429e51e788417fe9ff025e116b1
+    x-checker-data:
+      type: anitya
+      project-id: 6845
+      stable-only: true
+      url-template: https://boostorg.jfrog.io/artifactory/main/release/$version/source/boost_${major}_${minor}_$patch.tar.bz2
+buildsystem: simple
+build-commands:
+  - ./bootstrap.sh --prefix=/app
```
Comment (barthalion) [inline boost.yml:13]: Consider changing it to build&install only libraries which application uses.

File: edu.princeton.physics.WSJTX.metainfo.xml (hunk 1 above)
Comment (barthalion) [reply edu.princeton.physics.WSJTX.metainfo.xml:3]: Note it's better to add `<launchable type="desktop-id">edu.princeton.physics.WSJTX.desktop</launchable>` if you drop desktop suffix.

File: edu.princeton.physics.WSJTX.yml (hunk 2 above)
Comment (barthalion) [reply edu.princeton.physics.WSJTX.yml:26]: Yeah, flat structure for short-to-medium sized manifests is preferred.

File: edu.princeton.physics.WSJTX.yml [hunk 7]
```diff
@@ -0,0 +1,50 @@
+app-id: edu.princeton.physics.WSJTX
+runtime: org.kde.Platform
+runtime-version: '5.15'
+sdk: org.kde.Sdk
+command: wsjtx
+rename-desktop-file: wsjtx.desktop
+rename-icon: wsjtx_icon
+finish-args:
+  - --share=ipc
+  - --socket=x11
```
Comment (barthalion) [inline edu.princeton.physics.WSJTX.yml:10]:
```suggestion
  - --socket=fallback-x11
```

File: fftw3.yml (hunk 3 above)
Comment (AsciiWolf) [reply fftw3.yml:1]: Although it is a little outdated.

File: edu.princeton.physics.WSJTX.metainfo.xml (hunk 1 above)
Comment (AsciiWolf) [reply edu.princeton.physics.WSJTX.metainfo.xml:3]: It should not be needed if the desktop file name matches the app id.

File: fftw3.yml (hunk 3 above)
Comment (hfiguiere) [reply fftw3.yml:1]: merged. thanks. I was meaning to do it but got sidetracked.

File: edu.princeton.physics.WSJTX.yml (hunk 2 above)
Comment (xylo04) [reply edu.princeton.physics.WSJTX.yml:26]: Inlined

File: fftw3.yml (hunk 3 above)
Comment (xylo04) [reply fftw3.yml:1]: Now using FFTW3 from shared modules.

File: edu.princeton.physics.WSJTX.metainfo.xml (hunk 4 above)
Comment (xylo04) [reply edu.princeton.physics.WSJTX.metainfo.xml:9]: Used the first paragraph

File: edu.princeton.physics.WSJTX.metainfo.xml (hunk 5 above)
Comment (xylo04) [reply edu.princeton.physics.WSJTX.metainfo.xml:84]: Removed the dimensions

File: edu.princeton.physics.WSJTX.yml (hunk 7 above)
Comment (xylo04) [reply edu.princeton.physics.WSJTX.yml]: Done.

File: boost.yml (hunk 6 above)
Comment (xylo04) [reply boost.yml]: Sorry, I'm not familiar enough with Boost to understand this. What do I need to change?

File: edu.princeton.physics.WSJTX.metainfo.xml [hunk 8]
```diff
@@ -0,0 +1,33 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<component type="desktop">
+    <id>edu.princeton.physics.WSJTX</id>
+    <name>wsjtx</name>
```
Comment (AsciiWolf) [inline edu.princeton.physics.WSJTX.metainfo.xml:4]:
```suggestion
    <name>wsjtx</name>
    <provides>
        <id>wsjtx.desktop</id>
    </provides>
```

(So GNOME Software won't display this Flatpak as a separate entry, but as another source for the regular distribution package.)

File: boost.yml (hunk 6 above)
Comment (AsciiWolf) [reply boost.yml]: You can [take a look](https://www.google.com/search?q=site%3Agithub.com%2Fflathub+boost) what other Flatpaks use.

Comment (xylo04) [reply boost.yml]: This should give me a good idea
```
$ grep -rhi "#include <boost/" | sort | uniq
#include <boost/algorithm/string/predicate.hpp> 
#include <boost/config.hpp>                  
#include <boost/container/flat_map.hpp>                                                        
#include <boost/core/noncopyable.hpp>         
#include <boost/crc.hpp>                 
#include <boost/date_time/gregorian/greg_day.hpp>
#include <boost/date_time/posix_time/posix_time.hpp>
#include <boost/filesystem/fstream.hpp>
#include <boost/functional/hash.hpp>
#include <boost/lambda/lambda.hpp>             
#include <boost/lexical_cast.hpp>
#include <boost/log/attributes/clock.hpp>
#include <boost/log/attributes/counter.hpp>
#include <boost/log/attributes/current_process_id.hpp>
#include <boost/log/attributes/current_thread_id.hpp>
#include <boost/log/attributes.hpp>
#include <boost/log/attributes/mutable_constant.hpp>
#include <boost/log/common.hpp>
#include <boost/log/core.hpp>
#include <boost/log/expressions/formatters/date_time.hpp>
#include <boost/log/expressions.hpp>
#include <boost/log/expressions/keyword.hpp>
#include <boost/log/expressions/predicates/channel_severity_filter.hpp>
#include <boost/log/sinks/async_frontend.hpp>
#include <boost/log/sinks.hpp>
#include <boost/log/sinks/sync_frontend.hpp>
#include <boost/log/sinks/text_file_backend.hpp>
#include <boost/log/sinks/text_ostream_backend.hpp>
#include <boost/log/sources/global_logger_storage.hpp>
#include <boost/log/sources/record_ostream.hpp> 
#include <boost/log/sources/severity_channel_logger.hpp>
#include <boost/log/support/date_time.hpp>
#include <boost/log/trivial.hpp>
#include <boost/log/utility/exception_handler.hpp>
#include <boost/log/utility/manipulators/add_value.hpp>
#include <boost/log/utility/setup/console.hpp>
#include <boost/log/utility/setup/filter_parser.hpp>
#include <boost/log/utility/setup/from_stream.hpp>
#include <boost/log/utility/setup/settings.hpp> 
#include <boost/make_shared.hpp>
#include <boost/multi_index_container.hpp>
#include <boost/multi_index/hashed_index.hpp>
#include <boost/multi_index/key_extractors.hpp> 
#include <boost/multi_index/ordered_index.hpp>
#include <boost/range/iterator_range.hpp>
#include <boost/shared_ptr.hpp>
#include <boost/version.hpp>
```

Comment (xylo04) [reply boost.yml]: Done in 1e5643c

**Comment by xylo04** [comment]
bot, build edu.princeton.physics.WSJTX

**Comment by xylo04** [comment]
Screenshot from 2021-07-06 10-26-28
This build is decoding from audio well, but is not able to view serial TTYs to start a CAT protocol connection. I have the `device=all` permission, perhaps I need additional filesystem access? What does fldigi do...?

**Comment by AsciiWolf** [comment]
Yeah, the test build seems to work fine for me as well, although I did not test serial TTY.

wsjtx

One small thing that I have noticed is that by default, there are no audio devices selected. I had to manually turn them on in settings. But this may be a WSJT-X issue, not a Flatpak one.

**Comment by xylo04** [comment]
Serial TTY is working now, must have been a fluke or an ordering issue. I think I've seen in vanilla builds that the device list does not refresh when a new device is inserted.

@AsciiWolf, I think the issue of no audio device selected by default is in upstream as well.

Probably the biggest usability issue is that the RX and TX text fields are squashed, probably due to the QT version we're using. We can try to resolve that upstream; I wouldn't delay publishing this PR for that.

**Comment by AsciiWolf** [comment]
Here is a screenshot from RPM version of WSJT-X from Fedora 34:

wsjtx_noflatpak

These fields are squashed here too.

**Comment by AsciiWolf** [comment]
Oh, I just found that when I launch WSJT-X from a desktop menu, the fields are not squashed. This is because from the desktop file it launches WSJT-X like this: `wsjtx --style=fusion`

I have tried this launch option with the Flatpak version and it fixed the problem there as well. ;-)

**Comment by hfiguiere** [comment]
These are ALSA message, nothing to do with pipewire.

**Comment by AsciiWolf** [comment]
@xylo04 Here is a patch for the text fields issue:

```
diff --git a/edu.princeton.physics.WSJTX.yml b/edu.princeton.physics.WSJTX.yml
index 12568d2..7727c31 100644
--- a/edu.princeton.physics.WSJTX.yml
+++ b/edu.princeton.physics.WSJTX.yml
@@ -2,7 +2,7 @@ app-id: edu.princeton.physics.WSJTX
 runtime: org.kde.Platform
 runtime-version: '5.15'
 sdk: org.kde.Sdk
-command: wsjtx
+command: wsjtx.sh
 rename-desktop-file: wsjtx.desktop
 rename-icon: wsjtx_icon
 finish-args:
@@ -59,6 +59,10 @@ modules:
         x-checker-data:
           type: git
           tag-pattern: ^wsjtx-([\\d.]+)$
+      - type: script
+        dest-filename: wsjtx.sh
+        commands:
+          - wsjtx $@ --style=fusion
       - type: patch
         path: wsjtx-use-standard-icon-path.patch
       - type: file
@@ -68,6 +72,7 @@ modules:
       - -DWSJT_SKIP_MANPAGES=ON
       - -DWSJT_GENERATE_DOCS=OFF
     post-install:
+      - install -D wsjtx.sh /app/bin/wsjtx.sh
       - install -D edu.princeton.physics.WSJTX.metainfo.xml /app/share/metainfo/edu.princeton.physics.WSJTX.metainfo.xml
       - sed -i 's:Icon=wsjtx_icon:Icon=edu.princeton.physics.WSJTX:' /app/share/applications/message_aggregator.desktop
       - mv /app/share/applications/message_aggregator.desktop /app/share/applications/edu.princeton.physics.WSJTX.MessageAggregator.desktop
```

**Comment by xylo04** [comment]
bot, build edu.princeton.physics.WSJTX

**Comment by AsciiWolf** [comment]
bot, build edu.princeton.physics.WSJTX

**Comment by xylo04** [comment]
So far that's not helping me, I still get the +/- buttons instead of the smaller up/down ones.
image

**Comment by xylo04** [comment]
Running manually with that flag does fix the issue:
```
flatpak run edu.princeton.physics.WSJTX --style=fusion
```
image

**Comment by AsciiWolf** [comment]
Hmm, that's really strange. Are you sure you have the latest build installed?

**Comment by AsciiWolf** [comment]
Oops, I am dumb. I forgot that we have to change `Exec=` in the desktop file as well. I will send a patch.

**Comment by xylo04** [comment]
Sounds good, PR is fine. I didn't use the last one because I was already modifying in the area

**Comment by xylo04** [comment]
Looks like there's a weird issue with Pop_OS caching the desktop files. When I run from the Pop_OS (Gnome) launcher I don't get the fusion styling, but running from `flatpak run` I do get fusion styling.

**Comment by AsciiWolf** [comment]
@xylo04 When you run it from `flatpak run`, it directly runs the command specified in the Flatpak manifest. Here is a patch to fix the desktop file to use the same command:

```
diff --git a/edu.princeton.physics.WSJTX.yml b/edu.princeton.physics.WSJTX.yml
index 7727c31..eeaea32 100644
--- a/edu.princeton.physics.WSJTX.yml
+++ b/edu.princeton.physics.WSJTX.yml
@@ -74,5 +74,6 @@ modules:
     post-install:
       - install -D wsjtx.sh /app/bin/wsjtx.sh
       - install -D edu.princeton.physics.WSJTX.metainfo.xml /app/share/metainfo/edu.princeton.physics.WSJTX.metainfo.xml
+      - sed -i 's:Exec=wsjtx:Exec=wsjtx.sh:' /app/share/applications/wsjtx.desktop
       - sed -i 's:Icon=wsjtx_icon:Icon=edu.princeton.physics.WSJTX:' /app/share/applications/message_aggregator.desktop
       - mv /app/share/applications/message_aggregator.desktop /app/share/applications/edu.princeton.physics.WSJTX.MessageAggregator.desktop
```

**Comment by xylo04** [comment]
bot, build edu.princeton.physics.WSJTX

**Comment by xylo04** [comment]
Yes, that's fixed the desktop file.

**Comment by xylo04** [comment]
Several QSOs (radio contacts) made. Duplex audio, serial TTY, uploads to PSKreporter, UDP API and local log files are all working as expected.

**Pull request closed without merging**